### DISCUSSION TITLE
Installation Assistance

### DISCUSSION THREAD

**Discussion by lavarocker** [savonet/liquidsoap#2594] [Installation]
Hello.

I've tried several ways to install liquidsoap and keep running into errors.  Latest two attempts were:

1) From the liquidsoap book downloaded from the site.  It errored at `opam switch create 4.13.0`:
> "[ERROR] No compiler matching `4.13.0' found, use `opam switch list-available' to see what is available, or use `--packages' to select packages explicitly." 
When I run `opam switch list-available'  I get no matches found.

2)  Download and compile source code **recommended for newbs** per the site.  I downloaded liquidsoap-2.1.1bz2 file and when running compile got:
> "configure: error: version   "ocaml" {>= "4.12.0" & < "5.0.0"} or greater of the OCaml compiler is required to build liquidsoap"

My end goal is to get streaming working with icecast.

Is there a thorough and updated instruction set for complete newbs?

I'm using Ubuntu 22.04.1 LTS.

Thanks for any assistance provided.

**Comment by toots** [comment]
Hi @lavarocker. This seems weird and perhaps linked to an issue with your local `opam` install.

Have you tried the ubuntu jammy deb packages distributed with our release here: https://github.com/savonet/liquidsoap/releases/tag/v2.1.1 ?

**Reply by lavarocker** [reply]
Yes, that's what I had downloaded v21.1.1.  I'd like to reset and start from scratch as I've been trying to follow various instructions without success and certainly botched it up.  What is the recommended install procedure?  I need detailed instructions as I have not done much packaging/compiling work.

Thank you.

**Reply by sonyarianto** [reply]
@lavarocker 

I just do this when upgrade my LS from 2.1.0-1 to 2.1.1 and it works.

```
opam switch create 4.14.0
opam depext taglib mad lame vorbis cry samplerate ocurl liquidsoap fdkaac flac
opam install taglib mad lame vorbis cry samplerate ocurl liquidsoap fdkaac flac
```

of course above depencencies is for my stream requirements.

below is my opam list result

```
# Packages matching: installed
# Name                # Installed # Synopsis
angstrom              0.15.0      Parser combinators built for speed and memory-efficiency
base-bigarray         base
base-bytes            base        Bytes library distributed with the OCaml compiler
base-threads          base
base-unix             base
bigstringaf           0.9.0       Bigstring intrinsics and fast blits based on memcpy/memmove
camomile              1.0.2       A Unicode library
conf-lame             1           Virtual package relying on lame
conf-libcurl          2           Virtual package relying on a libcurl system installation
conf-libflac          1           Virtual package relying on libFLAC
conf-libogg           1           Virtual package relying on libogg
conf-libpcre          1           Virtual package relying on a libpcre system installation
conf-libvorbis        1           Virtual package relying on libvorbis
conf-mad              1           Virtual package relying on mad
conf-pkg-config       2           Check if pkg-config is installed and create an opam switch local pkgconfig folder
conf-samplerate       1           Virtual package relying on samplerate
conf-taglib           1           Virtual package relying on taglib
conf-which            1           Virtual package relying on which
cry                   0.6.7       OCaml client for the various icecast & shoutcast source protocols
csexp                 1.5.1       Parsing and printing of S-expressions in Canonical form
dtools                0.4.4       Library providing various helper functions to make daemons
dune                  3.4.1       Fast, portable, and opinionated build system
dune-configurator     3.4.1       Helper library for gathering system configuration
duppy                 0.9.2       Library providing monadic threads
fdkaac                0.3.2       Fraunhofer FDK AAC Codec Library
flac                  0.3.0       Bindings to libflac
gen                   1.0         Iterators for OCaml, both restartable and consumable
lame                  0.3.6       MP3 encoding library
liquidsoap            2.1.1       Swiss-army knife for multimedia streaming
mad                   0.5.2       Mad decoding library
menhir                20220210    An LR(1) parser generator
menhirLib             20220210    Runtime support library for parsers generated by Menhir
menhirSdk             20220210    Compile-time library for auxiliary tools related to Menhir
mm                    0.8.1       The mm library contains high-level to create and manipulate multimedia streams (audio, video, MIDI)
ocaml                 4.13.0      The OCaml compiler (virtual package)
ocaml-base-compiler   4.13.0      Official release 4.13.0
ocaml-compiler-libs   v0.12.4     OCaml compiler libraries repackaged
ocaml-config          2           OCaml Switch Configuration
ocaml-options-vanilla 1           Ensure that OCaml is compiled with no special options enabled
ocaml-syntax-shims    1.0.0       Backport new syntax to older OCaml versions
ocamlbuild            0.14.1      OCamlbuild is a build system with builtin rules to easily build most OCaml projects
ocamlfind             1.9.5       A library manager for OCaml
ocurl                 0.9.2       Bindings to libcurl
ogg                   0.7.3       Bindings to libogg
pcre                  7.5.0       Bindings to the Perl Compatibility Regular Expressions library
ppx_derivers          1.2.1       Shared [@@deriving] plugin registry
ppxlib                0.27.0      Standard library for ppx rewriters
result                1.5         Compatibility Result module
samplerate            0.1.6       Samplerate audio conversion library
sedlex                3.0         An OCaml lexer generator for Unicode
seq                   base        Compatibility package for OCaml's standard iterator type starting from 4.07.
sexplib0              v0.15.1     Library containing the definition of S-expressions and some base converters
stdlib-shims          0.3.0       Backport some of the new stdlib features to older compiler
stringext             1.6.0       Extra string functions for OCaml
taglib                0.3.9       Bindings for the taglib library
uchar                 0.0.2       Compatibility library for OCaml's Uchar module
uri                   4.2.0       An RFC3986 URI/URL parsing library
vorbis                0.8.0       Bindings to libvorbis
```

**Comment by vitoyucepi** [comment]
Hi, @lavarocker as for me opam is not the easiest way to install liquidsoap.
1. The easiest way is to use prebuilt packages.
   You can get them from assets of the release.
2. If you're familiar with docker, then you can use [official images](https://hub.docker.com/r/savonet/liquidsoap).
3. If you want to use opam, then this is the script I use
   ```sh
   sudo apt-get update
   sudo apt-get dist-upgrade
   sudo apt-get install curl make gcc patch unzip bubblewrap
   bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)"
   opam init --bare
   opam switch create 4.14.0
   eval $(opam env)
   opam install liquidsoap
   ```
   You'll probably want to install additional modules.
   To do that you could use `opam install` like this
   ```sh
   opam install ffmpeg lame vorbis 
   ```

**Reply by lavarocker** [reply]
Hi and thank for the above responses.

I tried both of the above suggestions and get the same "[ERROR] No compiler matching `4.1x.x' at the 'Opam switch create 4.1x.x'.

I'm hesitant to use Docker as unsure if that package configuration is as flexible as stand alone.

Maybe best to reset the test system and start over as I probably botched something important.

Still looking for a complete, comprehensive, and current guide for newbs if anyone has it.

Thanks.

**Reply by vitoyucepi** [reply]
Did you try the packages?
I've heard you're using Ubuntu:22.04.
Liquidsoap supports *amd64* and *arm64*.
1. Go to https://github.com/savonet/liquidsoap/releases/tag/v2.1.1.
2. Scroll down to the *Assets*.
3. Find *liquidsoap_2.1.1-ubuntu-jammy*
4. If your linux machine has *amd64* architecture, then copy the link to
   *liquidsoap_2.1.1-ubuntu-jammy-1_amd64.deb*
   If the architecture is *arm64*, then copy the link to
   *liquidsoap_2.1.1-ubuntu-jammy-1_arm64.deb*
5. Log into your machine by ssh.
6. Use wget to download the file with the link you copied previously.
   For *amd64* it is `wget https://github.com/savonet/liquidsoap/releases/download/v2.1.1/liquidsoap_2.1.1-ubuntu-jammy-1_amd64.deb`.
   It will save file to *liquidsoap_2.1.1-ubuntu-jammy-1_amd64.deb*.
7. Install package using `apt-get install ./liquidsoap_2.1.1-ubuntu-jammy-1_amd64.deb`

**Reply by lavarocker** [reply]
Yes I see; use the release specific version.  It does appear to have installed.  I see /usr/bin/liquidsoap exists in the system now.

Post install, following along in the book.pdf from the website I tried "liquidsoap 'output(sine())'" command and got the below so not sure it is working.

Also checking /var/log/liquidsoap the folder exists but there are no logs.

2022/09/05 18:36:14 >>> LOG START
2022/09/05 18:36:09 [main:3] Liquidsoap 2.1.1
2022/09/05 18:36:09 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] posix-time2=2.0.0 pcre=7.5.0 sedlex=3.0 menhirLib=20220210 curl=0.9.2 uri=4.2.0 memtrace=v0.2.2 mem_usage=0.0.4 dtools=0.4.4 duppy=0.9.2 cry=0.6.8 mm=0.8.1 xmlplaylist=0.1.5 lastfm=0.3.3 ogg=0.7.3 ogg.decoder=0.7.3 vorbis=0.8.1 vorbis.decoder=0.8.1 opus=0.2.2 opus.decoder=0.2.2 speex=0.4.1 speex.decoder=0.4.1 mad=0.5.2 flac=0.3.0 flac.ogg=0.3.0 flac.decoder=0.3.0 dynlink=[distributed with Ocaml] lame=0.3.6 shine=0.2.3 frei0r=0.1.2 fdkaac=0.3.2 theora=0.4.0 theora.decoder=0.4.0 ffmpeg=1.1.5 bjack=0.1.6 alsa=0.3.0 ao=0.2.4 samplerate=0.1.6 taglib=0.3.9 ssl=0.5.11 magic=0.7.3 camomile=1.0.2 inotify=2.3 faad=0.5.2 soundtouch=0.1.9 portaudio=0.2.3 pulseaudio=0.1.5 ladspa=0.2.2 dssi=0.1.5 tsdl=v0.9.9 tsdl-ttf=0.3.2 tsdl-image=0.3.2 camlimages=4.2.6 camlimages.freetype=5.0.4 cohttp-lwt-unix=5.0.0 prometheus-app=1.2 srt.constants=0.2.2 srt.types=0.2.2 srt.stubs=0.2.2 srt.stubs.locked=0.2.2 srt=0.2.2 lo=0.2.0 gd=1.0a5 irc-client-unix=[unspecified]
2022/09/05 18:36:09 [clock:3] Using native (high-precision) implementation for latency control
2022/09/05 18:36:14 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2022/09/05 18:36:14 [frame:3] Video frame size set to: 1280x720
2022/09/05 18:36:14 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2022/09/05 18:36:14 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2022/09/05 18:36:14 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2022/09/05 18:36:14 [sandbox:3] Sandboxing disabled
2022/09/05 18:36:14 [video.converter:3] Using preferred video converter: ffmpeg.
2022/09/05 18:36:14 [audio.converter:3] Using samplerate converter: libsamplerate.
2022/09/05 18:36:14 [clock.pulseaudio:2] **_Error when starting pulse_out(liquidsoap:): Pulseaudio error: Connection refused!_**
2022/09/05 18:36:14 [video.text:3] Using camlimages implementation
2022/09/05 18:36:14 [main:3] Shutdown started!
2022/09/05 18:36:14 [main:3] Waiting for main threads to terminate...
2022/09/05 18:36:14 [main:3] Main threads terminated.
2022/09/05 18:36:14 [threads:3] Shutting down scheduler...
2022/09/05 18:36:14 [threads:3] Scheduler shut down.
2022/09/05 18:36:14 [main:3] Cleaning downloaded files...
2022/09/05 18:36:14 [main:3] Freeing memory...

I reinstalled Pulseaudio but getting the same error.  I realize this is just a test but shouldn't it work?

Any ideas?

Thanks
